torch/utils/filelock.py

```diff
@@ -0,0 +1,37 @@
+from types import TracebackType
+from typing import Self
+
+from filelock import FileLock as base_FileLock
+
+from torch.monitor import _WaitCounter
+
+
+class FileLock(base_FileLock):
+    """
+    This behaves like a normal file lock.
+
+    However, it adds waitcounters for acquiring and releasing the filelock
+    as well as for the critical region within it.
+
+    pytorch.filelock.enter - While we're acquiring the filelock.
+    pytorch.filelock.region - While we're holding the filelock and doing work.
+    pytorch.filelock.exit - While we're releasing the filelock.
+    """
+
+    def __enter__(self) -> Self:
+        self.region_counter = _WaitCounter("pytorch.filelock.region").guard()
+        self.region_counter.__enter__()
+        with _WaitCounter("pytorch.filelock.enter").guard():
+            super().__enter__()
+        return self
```

Comment:
I don't think it makes a difference because I believe filelock returns self, but this is non-idiomatic: you should directly `return super().__enter__()` in case the `as` binding on context manager you are overriding is different.

Reply:
Huh - I guess I need to read the python inheritance docs again. I thought that if super().__enter__() returned self, it would return it's parent self, and thus we wouldn't end up calling the __exit__() below.

Reply:
Yep confirmed via tests, this behaves correctly.